benchmarking/benchmark.py

```diff
@@ -230,46 +235,71 @@ def run_experiments(config):
+        threshold = experiment['threshold']
+        size_a, size_b = get_exp_sizes(experiment)
+
+        for rep in range(repetition):
+            current_experiment = copy.deepcopy(experiment)
+            current_experiment['rep'] = rep + 1
+            logger.info('running experiment: {}'.format(current_experiment))
+            if repetition != 1:
+                logger.info('\trepetition {} out of {}'.format(rep + 1, repetition))
+            result = run_single_experiment(server, config, threshold, size_a, size_b, current_experiment)
+            results['experiments'].append(result)
+
     return results
 
 
+def run_single_experiment(server, config, threshold, size_a, size_b, experiment):
+    result = {}
+    credentials = {}
+    run = {}
+    try:
+        credentials = rest_client.project_create(server, config['schema'], 'mapping',
+                                                 "benchy_{}".format(experiment))
+        # upload clks
+        upload_binary_clks(config, size_a, size_b, credentials)
+        # create run
+        run = rest_client.run_create(server, credentials['project_id'], credentials['result_token'],
+                                     threshold,
+                                     "{}_{}".format(experiment, threshold))
+        # wait for result
+        run_id = run['run_id']
+        result['run_id'] = run_id
+        logger.info(f'waiting for run {run_id} to finish')
+        status = rest_client.wait_for_run(server, credentials['project_id'], run['run_id'],
+                                          credentials['result_token'], timeout=config['timeout'])
+        if status['state'] != 'completed':
+            raise RuntimeError('run did not finish!\n{}'.format(status))
+        logger.info('experiment successful. Evaluating results now...')
+        mapping = rest_client.run_get_result_text(server, credentials['project_id'], run['run_id'],
+                                                  credentials['result_token'])
+        mapping = json.loads(mapping)['mapping']
+        mapping = {int(k): int(v) for k, v in mapping.items()}
+        tt = score_mapping(mapping, *load_truth(config, size_a, size_b))
+        result.update(compose_result(status, tt, experiment, (size_a, size_b), threshold))
+    except Exception as e:
+        e_trace = format_exc()
+        logger.warning("experiment '{}' failed: {}".format(experiment, e_trace))
+        result.update({'name': experiment, 'status': 'ERROR', 'description': e_trace})
+    finally:
+        logger.info('cleaning up...')
+        delete_resources(config, credentials, run)
+
+    return result
+
+
+def push_result_s3(experiment_file):
+    client = boto3.client(
+        's3',
+        aws_access_key_id=os.getenv('AWS_ACCESS_KEY_ID'),
+        aws_secret_access_key=os.getenv('AWS_SECRET_ACCESS_KEY')
+    )
+    s3_bucket = "anonlink-benchmark-result"
+    client.upload_file(experiment_file, s3_bucket, "results.json")
```

Comment:
So if this benchmarking job *is* run multiple times in k8s with the same s3 bucket (extremely likely) the results get overridden. I suggest we include a timestamp or uuid in the uploaded filename.

Reply:
:+1: 
I will use timestamps, to more easily access them after (the job may not be kept on k8s, so the UUID may become hard to access).